sycl/doc/EnvironmentVariables.md

```diff
@@ -14,6 +14,7 @@ subject to change. Do not rely on these variables in production code.
 | SYCL_PI_TRACE | Described [below](#sycl_pi_trace-options)  | Enable specified level of tracing for PI. |
 | SYCL_BE | PI_OPENCL, PI_LEVEL_ZERO, PI_CUDA | Force SYCL RT to consider only devices of the specified backend during the device selection. |
 | SYCL_DEVICE_TYPE | One of: CPU, GPU, ACC, HOST | Force SYCL to use the specified device type. If unset, default selection rules are applied. If set to any unlisted value, this control has no effect. If the requested device type is not found, a `cl::sycl::runtime_error` exception is thrown. If a non-default device selector is used, a device must satisfy both the selector and this control to be chosen. This control only has effect on devices created with a selector. |
+| SYCL_DEVICE_FILTER (name TBD) | {backend:device_type:device_num} | Force SYCL RT to use the specified device with optional backend and device_num information. It can list any number of triples separated by commas. This env var affects all different device selectors and device discovery. When no device that can satisfy the triple by a default_selector, a heuristic will choose the device that has the closest match. Possible values of device_type are \*,host,cpu,gpu,acc. Possible values of backend are \*,opencl, level_zero, cuda. Device_num is an unsigned integer that indexes the enumeration of devices from sycl::platform::get_device() call. All triple entries are optional, but one of them should be present. For example, to use cpu and level_zero gpu device number 0, one can set SYCL_DEVICE_FILTER=cpu,level_zero:gpu::0. This environment variable will limit loading only specified plugins into the SYCL RT. For example, SYCL_DEVICE_FILTER=level_zero and cpu_selector will throw an error because the CPU device is not supported by the level_zero backend. |
```

Comment:
> When no device that can satisfy the triple by a default_selector, a heuristic will choose the device that has the closest match. 

Why is this special (different from other selectors) rule? What would be the closest match if, say, filter "cuda" is used on a machine with no CUDA HW?

Reply:
This is the same behavior as before. I just wanted clarify this default behavior because too many people asked "what if.." etc.
So, in your case, the cuda backend cannot be found so requesting a device from a default selector will return HOST device. If someone asks for a gpu_selector, an exception will be thrown.

Reply:
I think I disagree with that behavior.  The filters are meant to be mandatory not optional.  If you say CUDA and there are no CUDA devices, you should throw an error.

Reply:
Will that always be HOST if filter returns null? Can we just say that? Can we also give an exception for default_selector if filter returns null under **non-null** SYCL_DEVICE_FILTER?

Reply:
That's the default behavior for special selector such as gpu_selector.
The default selector always has a backup (HOST device) in any case.

Reply:
Yeah I guess it's just a question of what makes more sense.  I'm not overly concerned in any case - we just  need to document the behavior.

The spec does say the host device should always be available, but atm this is a non-standard extension anyway, so altering the behavior to throw an error is not unreasonable.  If the host fallback behavior is desired, the user can always specify host as an optional filter.

Reply:
> That's the default behavior for special selector such as gpu_selector.
You mean "default_selector" right? I am with @jbrodman to get rid of this special behavior if SYCL_DEVICE_FILTER is used

Reply:
> > That's the default behavior for special selector such as gpu_selector.
> 
> You mean "default_selector" right? I am with @jbrodman to get rid of this special behavior if SYCL_DEVICE_FILTER is used

Actually not. I was replying to his comment that we should throw an exception no matter which device selector is used.
In fact, the current behavior is written in the spec, and I am not planning to change the current behavior of the existing device selectors. That's what we discussed through several meetings.

Reply:
Like I said - I don't care enough for it to be a blocker now, but please clearly document when this does and does not apply to the default_selector.

Reply:
> Like I said - I don't care enough for it to be a blocker now, but please clearly document when this does and does not apply to the default_selector.

Oops, I just realized that I mistakenly removed that sentence. It was there before, but I think I mistakenly removed that sentence. I will put it back. Sorry for my confusion.

Reply:
Done